packages/itwinui-react/src/core/utils/providers/HydrationProvider.tsx

```diff
@@ -0,0 +1,67 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import * as React from 'react';
+import { useSyncExternalStore, useIsClient } from '../hooks/index.js';
+
+const HydrationContext = React.createContext(false);
+const noopSubscribe = () => () => {};
+const isServer = typeof window === 'undefined';
+
+/**
+ * Hook that returns the hydration status of the app.
+ *
+ * @returns one of the following values:
+ * - `"hydrated"` after hydration is *definitely* complete (or is a pure client render)
+ * - `"hydrating"` if we know for sure that hydration is happening (in React 18)
+ * - `undefined` if the hydration status is unknown
+ *
+ * @private
+ */
+export const useHydration = () => {
+  // Returns true only if getServerSnapshot is called on the client.
+  // In React 18, this is true during hydration.
+  const hydrating = useSyncExternalStore(
+    noopSubscribe,
+    () => false,
+    () => !isServer,
+  );
```

Comment:
I'm a little confused about why we're using `useSyncExternalStore` here.

Reply:
`useSyncExternalStore` is the only way to check for hydration in react.

- `noopSubscribe` is just an empty function. It needs to be referentially stable, so it's declared outside of the component.
- `getSnapshot` returns false (called during non-hydration client renders)
- `getServerSnapshot` returns true on the server and during hydration. So I'm using a boolean check to make it only return true during hydration.